### PR TITLE
[FIX] mail: update email_from on template change

### PR DESCRIPTION
**Pull request by nd-dew**
Previously, when switching from an email template that overrides the email_from field to one that does not, the original email_from value was mistakenly passed as an argument. Consequently, emails were sent with the wrong sender, using the address from the template instead. Whereas, it is expected behavior that when a template does not specify any user, the current user's email address should be used as the sender.

[Changes introduced by this commit]
This commit addresses the issue by updating the sender email to the current user's email address when the template being updated to does not override the sender email field.

[Reproduce]
- Install -i contacts
- Log in as user U
- Create 2 Email templates that "Applies to" Contact (Settings/Technical/Email/Email Templates):
	- T1: set "From" field in "Email Configuration" tab (different than U's email)
	- T2: without set "From" field
- Go to a contact app, select any user
- Click "Send message" -> Expand "Full Composer"
- Select template T1 (with set email)
- Select template T2 (without set email) and Send
- BUG: email gets send with email from T1 (even tho we selected T2, without any email)

opw-3769188

# Encouraging discussion
I would like to encourage discussion on the fix, as I can imagine different approaches to this problem. (Like maybe instead we could not sent the `sent_from` argument from the fronted side and relay on the mechanism of default values). Any comments highly welcomed

### CODE REVIEW

**Review by nle-odoo — APPROVED** [review]
if it's what we want looks ok

it would make more sense to me to change it in the mail.compose.message()_onchange_template_id. because:

- eg, now it seems that if people showed the email_from field, they will get the previous value until it's modified)
- if we remove the template rather than setting a template without email_from, the behavior will be different:

    - set template without email_from
    - set email_from to caramail@example.com
    - set to another template without email_from
    => the email_from is reset to the current user email

    - set template without email_from
    - set email_from to caramail@example.com
    - remove email template
    => the email_from is kept

from what I remember, we wanted to only change the current value if the value was set in the current template, but here I guess it could make sense to have a different behavior if by default the user can't see or edit the email_from field (maybe you can be I've not seen how)

side note: mail code owners will probably want a test

**Comment by nd-dew** [comment]
Thanks for having a look @nle-odoo 
If I understood correctly, you proposed to move the fix to the function [def _onchange_template_id(...)](https://github.com/odoo/odoo/blob/cd194bd4dddd4fb8e2e1f8262b1f7f3f90a1cead/addons/mail/wizard/mail_compose_message.py#L561) to handle a situation where template_id is unset.

That is true that that case is not handled by my fix, however in this situation `email_from` is assigned to a default value in here:
https://github.com/odoo/odoo/blob/cd194bd4dddd4fb8e2e1f8262b1f7f3f90a1cead/addons/mail/wizard/mail_compose_message.py#L56-L63 So I think that desired behavior is achieved

Let me do a quick demo
https://github.com/odoo/odoo/assets/33809926/944162b3-1bde-4705-b7c2-dfe807b3522b




> from what I remember, we wanted to only change the current value if the value was set in the current template, but here I guess it could make sense to have a different behavior if by default the user can't see or edit the email_from field (maybe you can be I've not seen how)

not sure how to understand that

> side note: mail code owners will probably want a test

Thanks for the reminder

**Comment by nd-dew** [comment]
added test

**Comment by nd-dew** [comment]
rebased

**Comment by reth-odoo** [comment]
Maybe I'm missing something from the previous comments here but shouldn't the fix just be in the onchange of the composer? Doesn't seem right to change the behavior of template value generation to fix an issue with the composer.

Can't we do "if there was a template before, and the template has a value for X, and the current value for X in the composer is the same, then and only then reset the field"? Without having to touch templates to somehow force them to use a value it wouldn't have used. As that method is used by other things than the composer and presumably by a decent amount of custom code out there.

Let me know if that was already addressed though ^^

**Comment by nd-dew** [comment]
## Fix complete rework
Taking new comments into account, I've tried implementing the following logic:
image

### ¿All fields?
I've only included fields updated the template, I infer that is what was meant by "all fields"

### Fix location
Also I see that this fix can be located both in
-  `_onchange_template_id` 
-  `_onchange_template_id_wrapper` 
I've put it in the former to keep it specyfic to situation where template _id is spedifided. Not 100% sure if that the right approach. 

https://github.com/odoo/odoo/blob/a66690ef6d196d3ecd9a3fc51c7f8aec791d1230/addons/mail/wizard/mail_compose_message.py#L575-L578

**Comment by nd-dew** [comment]
## Handle case: composer doesn't have some fields from template
In composer doesn't have some field -> grab default  (i.e. don't remove this field from the ones that will be set to default)
```diff
             if fields_to_reset_to_default:
                 values.update(self.default_get(fields_to_reset_to_default))
                 for field in fields_to_reset_to_default:
-                    if self._origin.template_id[field] != self[field]:
+                    if field in self and self._origin.template_id[field] != self[field]:
                         fields_to_reset_to_default -= field
                 self.with_context(default_composition_mode=composition_mode, default_model=model, default_res_id=res_id).default_get(fields_to_reset_to_default)
         else:
```

File: addons/mail/wizard/mail_compose_message.py
Comment (nd-dew) [inline addons/mail/wizard/mail_compose_message.py:605]: In composer doesn't have some field -> grab default  (i.e. don't remove this field from the ones that will be set to default)

on a second thought, maybe we should just leave it as it was (eg. if "email_from" wasn't set, leave it unset)

**Comment by nd-dew** [comment]
# Another smaller rework
So concept mentioned in https://github.com/odoo/odoo/pull/158000#issuecomment-2020122327 isn't feasible to realinize in stable branch since we can't get `self._origin.template_id` (values before change) in the onchange

## After consultation with the PO
 we'll follow the following conception:
> "You selected template, so you get the template values"

or
> "If you change the template, all the values a template can set are set, I don't care that you had already typed your subject

This way we'll place more responsibility on a user, but we'll get consistent behavior.

File: addons/test_mail/tests/test_mail_composer.py
Comment (reth-odoo) [inline addons/test_mail/tests/test_mail_composer.py:462]: Not sure why this doesn't work but you could use a `Form(composer)` to get a "front-end-like" api which is more relevant for the fixed use case anyway.

**Review by reth-odoo — COMMENTED** [review]
Doesn't seem to fix it for "mass mode" ^^
That's the previous condition, and it's easily reachable from the same UI (just select multiple records from the list view of contacts, then action -> send mail)

Similar fix should be applicable, maybe putting some logic after the `elif` blocks?

**Comment by nd-dew** [comment]
Commit above :arrow_up: 
## Mass mode

I moved the fix into a separate if block, so it has a chance of capturing both **mass mode** and **comment**. Also some refactoring so the code is much more concise, may have overdid it :balance_scale:

File: addons/test_mail/tests/test_mail_composer.py
Comment (nd-dew) [reply addons/test_mail/tests/test_mail_composer.py:462]: > Not sure why this doesn't work

I'm not sure what do you mean? What isn't working in that part of the test?

PS: I would like to expand this test to check also mass_mailing mode probably

**Comment by nd-dew** [comment]
Commit above ⬆️: style fixes & rebased

File: addons/test_mail/tests/test_mail_composer.py
Comment (reth-odoo) [reply addons/test_mail/tests/test_mail_composer.py:462]: Sorry, I meant "I don't know why onchange isn't triggered" basically. Normally you want to avoid calling it manually as it hides the issue that it's not being triggered correctly.

File: addons/mail/wizard/mail_compose_message.py
Comment (reth-odoo) [inline addons/mail/wizard/mail_compose_message.py:601]: As we're not getting those (not in `fields_to_reset`), no need to force their default in context

**Review by reth-odoo — COMMENTED** [review]
Might be missing something but it doesn't seem to work functionally, when I set a template without a subject, it doesn't reset the subject.

**Comment by nd-dew** [comment]
Hi @reth-odoo, I changed the test to use **From**.

## Subject field not being reset
With current fix, when I switch to a template without an `email_from`, my fix will notice it and call `default_get` which will get me the current user `email_from`. This happens here
https://github.com/odoo/odoo/blob/2c55aa8eeda7e2629674f3216767ee684e42501e/addons/mail/wizard/mail_compose_message.py#L43-L83

Now when I switch from template_with_subject to a one without it. My fix will detect it and try to obtain a default value for subject (sketch below), however since subject doesn't have default value and there is no mechanism to set it to sth inside of [MailComposer.defulat_get()](https://github.com/odoo/odoo/blob/2c55aa8eeda7e2629674f3216767ee684e42501e/addons/mail/wizard/mail_compose_message.py#L43)  we stay with the current template subject. Which I think is fine unless I missed something.

image

We could add a default value to subject field, that would set it to empty, but I don't know if that's a sensible thing to do.
```diff
diff --git a/addons/mail/wizard/mail_compose_message.py b/addons/mail/wizard/mail_compose_message.py
index 82f9b34ca100..0a0b2df7817d 100644
--- a/addons/mail/wizard/mail_compose_message.py
+++ b/addons/mail/wizard/mail_compose_message.py
@@ -89,7 +89,7 @@ class MailComposer(models.TransientModel):
         ])
 
     # content
-    subject = fields.Char('Subject', compute=False)
+    subject = fields.Char('Subject', compute=False, default="")
```

check out the push below :arrow_down:

File: addons/mail/wizard/mail_compose_message.py
Comment (nd-dew) [reply addons/mail/wizard/mail_compose_message.py:601]: so I don't really need to update context right?

```diff
-                values.update(self.with_context(**context_dict).default_get(fields_to_reset))
+                values.update(self.default_get(fields_to_reset))
```

**Comment by nd-dew** [comment]
## Mass mode test
Forgot to include mass mode in the test before so here it is :arrow_down: 

## no context update
if I understood correctly, I don't need to update context in there, removing it from my fix :arrow_down:

**Review by reth-odoo — COMMENTED** [review]
> We could add a default value to subject field, that would set it to empty, but I don't know if that's a sensible thing to do.

I wouldn't do that, it just adds noise to field handling just to make it convenient for this specific case. I would just add something to the code so if default_get returns nothing for a value, we set it to False instead. Then let the ORM handle it however that field type handles `False`
```py
reset_values =  {field: False for field in fields_to_reset}
reset_values.update(self.default_get(fields_to_reset))
```

**Comment by nd-dew** [comment]
@reth-odoo have a look:
- I modified the test so it performs all actions for both 
- I used your approach in with the `reset_values` in the fix

it passes the test just fine, but it (updating from subject to no subject) won't work if you try doing that from frontend. Why? Cause the subject default value is passed by in the context so the BaseModel.default_get() will grab and set it here
https://github.com/odoo/odoo/blob/2331e69701c5dc468d4fe5f9d800aa325d67469d/odoo/models.py#L1409

File: addons/test_mail/tests/test_mail_composer.py
Comment (reth-odoo) [inline addons/test_mail/tests/test_mail_composer.py:491]: Maybe you can keep those and make it explicit "roll back" is not supported for body/subject.

**Review by reth-odoo — COMMENTED** [review]
> default subject is set

Indeed, it seems like `context="{'default_model': model, 'default_body_html': body, 'default_subject': subject}"` in the template completely overwrites the current context in the front-end and we can't ever get back the original default values...

It can work if you remove that line and do a bunch of other things to make it recompute defaults properly. But that's not super stable friendly (if the view isn't updated, it won't work) and overall it adds a lot of weird stuff to satisfy the logic of `default_get`/`get_record_data`

Also not worth "fixing" in the front-end, assuming that's even possible.

I guess it's fine if it doesn't work on subject/body, let's just fix this for the email fields which are the most annoying ones anyway. Would just ignore subject and body from the "fields to reset" then.

I think this should be the last update, thanks for the followups!

File: addons/test_mail/tests/test_mail_composer.py
Comment (nd-dew) [reply addons/test_mail/tests/test_mail_composer.py]: I just narrowed down the fix to handle only the email_from and reply_to fields. However, I'm unsure how to tweak the test to reflect the behavior of body and subject, as it would need context insertion. Yet, relying on frontend context might lead to irrelevant tests if the frontend changes its behavior later.

Comment (reth-odoo) [reply addons/test_mail/tests/test_mail_composer.py]: I meant you could just check that setting a template set the subject, but setting another template without subject keeps the same subject. Just as a documentation step.

But not testing it is fine actually, it doesn't really matter what we do with it.

**Review by reth-odoo — APPROVED** [review]
Seems like test might need an update otherwise lgtm, thanks for the fix

File: addons/test_mail/tests/test_mail_composer.py
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_composer.py:457]: There is a ``TestComposerForm`` class in this file for tests that more specifically use the Form tool to check what happens to the composer values ... Could be moved in that section :) (and there are a lot of additional tests in next version in that class so it makes sense to add this one there I think)

File: addons/test_mail/tests/test_mail_composer.py
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_composer.py:464]: :socks: Try to be consistent with quotes usage, use double quotes everywhere (personally I don't care but other people are becoming more strict)
```suggestion
            "email_from": "not_current_user@template_with_sender_subject.com",
```

File: addons/test_mail/tests/test_mail_composer.py
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_composer.py:470]: :socks: 
```suggestion
        for composition_mode in ["comment", "mass_mail"]:
```

File: addons/test_mail/tests/test_mail_composer.py
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_composer.py:471]: as msg will be the same for each run, it does not add much value :) purpose is to give values that distinguish the test (as those are outputed in the logs), so would say only composition_mode should be used as subTest parameters

File: addons/test_mail/tests/test_mail_composer.py
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_composer.py:463]: Maybe, if someday we add some more variations (more fields to check), we could use a more generic naming like
```suggestion
        template_complete = self.template.copy({
```

File: addons/test_mail/tests/test_mail_composer.py
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_composer.py:460]: Note that I would test reply_to reset, as it is part of this patch :)

**Review by tde-banana-odoo — COMMENTED** [review]
Thanks you both for your work :) added some nitpicking remarks while passing by :)

File: addons/test_mail/tests/test_mail_composer.py
Comment (nd-dew) [reply addons/test_mail/tests/test_mail_composer.py:471]: thx for the tip, appreciate it

File: addons/test_mail/tests/test_mail_composer.py
Comment (nd-dew) [reply addons/test_mail/tests/test_mail_composer.py:460]: Good catch! I checked again on `reply_to` behavior and it seem correct without my fix, which isn't doing anything for "reply_to" as it doesn't have any specific logic in the [`MailComposer.default_get`](https://github.com/odoo/odoo/blob/a161e7a2dd46ee1b25cc1b64161fc43864b007cb/addons/mail/wizard/mail_compose_message.py#L43) nor does it have any default value. 

-> simplifying to email_from only 

image

File: addons/test_mail/tests/test_mail_composer.py
Comment (nd-dew) [reply addons/test_mail/tests/test_mail_composer.py:457]: sure thing, moving to TestComposerForm

**Comment by nd-dew** [comment]
@reth-odoo can we merge it ?

File: addons/test_mail/tests/test_mail_composer.py
Comment (reth-odoo) [inline addons/test_mail/tests/test_mail_composer.py:486]: Wild new line

File: addons/test_mail/tests/test_mail_composer.py
Comment (reth-odoo) [inline addons/test_mail/tests/test_mail_composer.py:222]:
```suggestion
        template_no_subject = template_no_sender.copy({"subject": ""})
```

File: addons/test_mail/tests/test_mail_composer.py
Comment (reth-odoo) [inline addons/test_mail/tests/test_mail_composer.py:220]: This lint has practical value in allowing for 'cleaner' diffs if we need to update the parameters
```suggestion
            "subject": "subject: template_complete",
            })
```

**Review by reth-odoo — APPROVED** [review]
Yep, just a few linting nitpicks otherwise still seems fine to me :+1:

**Comment by nle-odoo** [comment]
robodoo delegate+

File: addons/mail/wizard/mail_compose_message.py
Comment (tde-banana-odoo) [inline addons/mail/wizard/mail_compose_message.py:599]: Could be simplified, isn't it ?
```suggestion
            if 'email_from' not in values:
```

File: addons/mail/wizard/mail_compose_message.py
Comment (tde-banana-odoo) [inline addons/mail/wizard/mail_compose_message.py:600]: Maybe we should ensure a bad default_get implementation (aka not respecting fields) does not update other fields, like
```suggestion
                values['email_from'] = self.default_get('email_from').get('email_from')
```

File: addons/test_mail/tests/test_mail_composer.py
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_composer.py:226]: I feel like you don't use a linter as you regularly have small linting issues, do you ? In 15 the ci/style is not running but you will end up with a red flag for 16.0... so check your linter (or setup one if you don't).
```suggestion
        for composition_mode in ["comment", "mass_mail"]:
```

**Review by tde-banana-odoo — COMMENTED** [review]
Zboing !

**Comment by nd-dew** [comment]
@robodoo r+

File: addons/mail/wizard/mail_compose_message.py
Comment (nd-dew) [reply addons/mail/wizard/mail_compose_message.py:600]: Thanks @SButko! Good catch!
Since this one is merged, I'm fixing it here. https://github.com/odoo/odoo/pull/166319

PS. how come you stumbled upon this PR ;)

Comment (tde-banana-odoo) [reply addons/mail/wizard/mail_compose_message.py:600]: Always wait for final approval before r+'ing, could have easily been avoided ^^

Comment (SButko) [reply addons/mail/wizard/mail_compose_message.py:600]: @nd-dew We have a customization where we modify the behavior of `default_get` and our tests started to fail because we expect to get list of fields. After some investigation, found this PR which caused an issue for us :)
Thanks for the fix!

**Pull request closed without merging**
